diff -urN mozilla-2.0/bluegriffon/app/profile/bluegriffon-prefs.js mozilla-2.0-patched/bluegriffon/app/profile/bluegriffon-prefs.js
--- mozilla-2.0/bluegriffon/app/profile/bluegriffon-prefs.js	2013-05-31 01:11:19.000000000 +1100
+++ mozilla-2.0-patched/bluegriffon/app/profile/bluegriffon-prefs.js	2013-06-29 11:08:00.638597978 +1100
@@ -125,8 +125,8 @@
 pref("bluegriffon.structurebar.role.show", true);
 
 // updates
-pref("bluegriffon.updates.check.enabled", true);
-pref("bluegriffon.updates.frequency", "launch");
+pref("bluegriffon.updates.check.enabled", false);
+pref("bluegriffon.updates.frequency", "manually");
 
 pref("html5.enable", true);
 pref("bluegriffon.defaults.doctype", "kHTML5");
