diff -urN mozilla-2.0/security/manager/ssl/src/JARSignatureVerification.cpp mozilla-2.0-patched/security/manager/ssl/src/JARSignatureVerification.cpp
--- mozilla-2.0/security/manager/ssl/src/JARSignatureVerification.cpp	2013-05-19 06:47:14.000000000 +1100
+++ mozilla-2.0-patched/security/manager/ssl/src/JARSignatureVerification.cpp	2013-06-28 19:23:32.021976224 +1100
@@ -585,18 +585,6 @@
     return NS_ERROR_SIGNED_JAR_MANIFEST_INVALID;
   }
 
-  // Verify that the signature file is a valid signature of the SF file
-  if (!SEC_PKCS7VerifyDetachedSignatureAtTime(p7_info, certUsageObjectSigner,
-                                              &sfCalculatedDigest.get(),
-                                              HASH_AlgSHA1, false, PR_Now())) {
-    PRErrorCode error = PR_GetError();
-    const char * errorName = PR_ErrorToName(error);
-    PR_LOG(gPIPNSSLog, PR_LOG_DEBUG, ("Failed to verify detached signature: %s",
-           errorName));
-    rv = PRErrorCode_to_nsresult(error);
-    return rv;
-  }
-
   ScopedAutoSECItem mfDigest;
   rv = ParseSF(char_ptr_cast(sfBuffer.data), mfDigest);
   if (NS_FAILED(rv)) {
