diff -uNrp mozilla-2.0.orig/config/rules.mk mozilla-2.0/config/rules.mk
--- mozilla-2.0.orig/config/rules.mk	2012-07-16 17:43:52.000000000 +0200
+++ mozilla-2.0/config/rules.mk	2012-07-16 19:23:11.054304420 +0200
@@ -1452,7 +1452,7 @@ $(XPIDL_GEN_DIR)/%.h: %.idl $(XPIDL_DEPS
 	$(PYTHON_PATH) \
 	  $(PLY_INCLUDE) \
 	  -I$(topsrcdir)/xpcom/idl-parser \
-	  $(topsrcdir)/xpcom/idl-parser/header.py --cachedir=$(DEPTH)/xpcom/idl-parser $(XPIDL_FLAGS) $(_VPATH_SRCS) -d $(MDDEPDIR)/$(@F).pp -o $@
+	  $(topsrcdir)/xpcom/idl-parser/header.py --cachedir=$(DEPTH)/xpcom/idl-parser/cache $(XPIDL_FLAGS) $(_VPATH_SRCS) -d $(MDDEPDIR)/$(@F).pp -o $@
 	@if test -n "$(findstring $*.h, $(EXPORTS))"; \
 	  then echo "*** WARNING: file $*.h generated from $*.idl overrides $(srcdir)/$*.h"; else true; fi
 
@@ -1465,7 +1465,7 @@ $(XPIDL_GEN_DIR)/%.xpt: %.idl $(XPIDL_DE
 	  $(PLY_INCLUDE) \
 	  -I$(topsrcdir)/xpcom/idl-parser \
 	  -I$(topsrcdir)/xpcom/typelib/xpt/tools \
-	  $(topsrcdir)/xpcom/idl-parser/typelib.py --cachedir=$(DEPTH)/xpcom/idl-parser $(XPIDL_FLAGS) $(_VPATH_SRCS) -d $(MDDEPDIR)/$(@F).pp -o $@
+	  $(topsrcdir)/xpcom/idl-parser/typelib.py --cachedir=$(DEPTH)/xpcom/idl-parser/cache $(XPIDL_FLAGS) $(_VPATH_SRCS) -d $(MDDEPDIR)/$(@F).pp -o $@
 
 # no need to link together if XPIDLSRCS contains only XPIDL_MODULE
 ifneq ($(XPIDL_MODULE).idl,$(strip $(XPIDLSRCS)))
diff -uNrp mozilla-2.0.orig/js/src/config/rules.mk mozilla-2.0/js/src/config/rules.mk
--- mozilla-2.0.orig/js/src/config/rules.mk	2012-07-16 17:43:55.000000000 +0200
+++ mozilla-2.0/js/src/config/rules.mk	2012-07-16 19:22:49.085408149 +0200
@@ -1452,7 +1452,7 @@ $(XPIDL_GEN_DIR)/%.h: %.idl $(XPIDL_DEPS
 	$(PYTHON_PATH) \
 	  $(PLY_INCLUDE) \
 	  -I$(topsrcdir)/xpcom/idl-parser \
-	  $(topsrcdir)/xpcom/idl-parser/header.py --cachedir=$(DEPTH)/xpcom/idl-parser $(XPIDL_FLAGS) $(_VPATH_SRCS) -d $(MDDEPDIR)/$(@F).pp -o $@
+	  $(topsrcdir)/xpcom/idl-parser/header.py --cachedir=$(DEPTH)/xpcom/idl-parser/cache $(XPIDL_FLAGS) $(_VPATH_SRCS) -d $(MDDEPDIR)/$(@F).pp -o $@
 	@if test -n "$(findstring $*.h, $(EXPORTS))"; \
 	  then echo "*** WARNING: file $*.h generated from $*.idl overrides $(srcdir)/$*.h"; else true; fi
 
@@ -1465,7 +1465,7 @@ $(XPIDL_GEN_DIR)/%.xpt: %.idl $(XPIDL_DE
 	  $(PLY_INCLUDE) \
 	  -I$(topsrcdir)/xpcom/idl-parser \
 	  -I$(topsrcdir)/xpcom/typelib/xpt/tools \
-	  $(topsrcdir)/xpcom/idl-parser/typelib.py --cachedir=$(DEPTH)/xpcom/idl-parser $(XPIDL_FLAGS) $(_VPATH_SRCS) -d $(MDDEPDIR)/$(@F).pp -o $@
+	  $(topsrcdir)/xpcom/idl-parser/typelib.py --cachedir=$(DEPTH)/xpcom/idl-parser/cache $(XPIDL_FLAGS) $(_VPATH_SRCS) -d $(MDDEPDIR)/$(@F).pp -o $@
 
 # no need to link together if XPIDLSRCS contains only XPIDL_MODULE
 ifneq ($(XPIDL_MODULE).idl,$(strip $(XPIDLSRCS)))
diff -uNrp mozilla-2.0.orig/js/xpconnect/src/Makefile.in mozilla-2.0/js/xpconnect/src/Makefile.in
--- mozilla-2.0.orig/js/xpconnect/src/Makefile.in	2012-07-16 17:43:56.000000000 +0200
+++ mozilla-2.0/js/xpconnect/src/Makefile.in	2012-07-16 19:22:15.211568090 +0200
@@ -155,7 +155,7 @@ dom_quickstubs.cpp: $(srcdir)/dom_quicks
 	  -I$(topsrcdir)/xpcom/idl-parser \
 	  $(srcdir)/qsgen.py \
 	  --idlpath=$(DEPTH)/dist/idl \
-	  --cachedir=$(DEPTH)/xpcom/idl-parser \
+	  --cachedir=$(DEPTH)/xpcom/idl-parser/cache \
 	  --header-output dom_quickstubs.h \
 	  --stub-output dom_quickstubs.cpp \
 	  --makedepend-output $(MDDEPDIR)/dom_qsgen.pp \
@@ -176,7 +176,7 @@ dombindings_gen.h: $(srcdir)/dombindings
 	  -I$(topsrcdir)/xpcom/idl-parser \
 	  $(srcdir)/dombindingsgen.py \
 	  --idlpath=$(DEPTH)/dist/idl \
-	  --cachedir=$(DEPTH)/xpcom/idl-parser \
+	  --cachedir=$(DEPTH)/xpcom/idl-parser/cache \
 	  --header-output dombindings_gen.h \
 	  $(srcdir)/dombindings.conf
 
@@ -193,7 +193,7 @@ dombindings_gen.cpp: $(srcdir)/dombindin
 	  -I$(topsrcdir)/xpcom/idl-parser \
 	  $(srcdir)/dombindingsgen.py \
 	  --idlpath=$(DEPTH)/dist/idl \
-	  --cachedir=$(DEPTH)/xpcom/idl-parser \
+	  --cachedir=$(DEPTH)/xpcom/idl-parser/cache \
 	  --stub-output dombindings_gen.cpp \
 	  --makedepend-output $(MDDEPDIR)/dombindingsgen.pp \
 	  $(srcdir)/dombindings.conf
@@ -212,7 +212,7 @@ DictionaryHelpers.h: $(srcdir)/dictionar
 	  -I$(topsrcdir)/xpcom/idl-parser \
 	  $(srcdir)/dictionary_helper_gen.py \
 	  -I $(DEPTH)/dist/idl \
-	  --cachedir=$(DEPTH)/xpcom/idl-parser \
+	  --cachedir=$(DEPTH)/xpcom/idl-parser/cache \
 	  --header-output DictionaryHelpers.h \
 	  $(srcdir)/dictionary_helper_gen.conf
 
@@ -227,7 +227,7 @@ DictionaryHelpers.cpp: $(srcdir)/diction
 	  -I$(topsrcdir)/xpcom/idl-parser \
 	  $(srcdir)/dictionary_helper_gen.py \
 	  -I $(DEPTH)/dist/idl \
-	  --cachedir=$(DEPTH)/xpcom/idl-parser \
+	  --cachedir=$(DEPTH)/xpcom/idl-parser/cache \
 	  --header-output DictionaryHelpers.h \
 	  --stub-output DictionaryHelpers.cpp \
 	  --makedepend-output $(MDDEPDIR)/dictionary_helper_gen.pp \
diff -uNrp mozilla-2.0.orig/xpcom/idl-parser/Makefile.in mozilla-2.0/xpcom/idl-parser/Makefile.in
--- mozilla-2.0.orig/xpcom/idl-parser/Makefile.in	2012-07-16 17:44:00.000000000 +0200
+++ mozilla-2.0/xpcom/idl-parser/Makefile.in	2012-07-16 19:20:31.966055578 +0200
@@ -61,7 +61,7 @@ export:: $(PARSER_SRCS) $(PLY_PROGS)
 	$(PYTHON_PATH) \
 	  $(PLY_INCLUDE) \
 	  -I$(topsrcdir)/xpcom/idl-parser \
-	  $(topsrcdir)/xpcom/idl-parser/header.py --cachedir=. --regen
+	  $(topsrcdir)/xpcom/idl-parser/header.py --cachedir=$(DEPTH)/xpcom/idl-parser/cache --regen
 
 check::
 	$(PYTHON_PATH) \
diff -uNrp mozilla-2.0.orig/xpcom/typelib/xpidl/Makefile.in mozilla-2.0/xpcom/typelib/xpidl/Makefile.in
--- mozilla-2.0.orig/xpcom/typelib/xpidl/Makefile.in	2012-07-16 17:44:00.000000000 +0200
+++ mozilla-2.0/xpcom/typelib/xpidl/Makefile.in	2012-07-16 19:21:19.258832276 +0200
@@ -48,8 +48,8 @@ SDK_BINARY = \
   $(topsrcdir)/xpcom/idl-parser/xpidl.py \
   $(topsrcdir)/xpcom/idl-parser/header.py \
   $(topsrcdir)/xpcom/idl-parser/typelib.py \
-  $(DEPTH)/xpcom/idl-parser/xpidllex.py \
-  $(DEPTH)/xpcom/idl-parser/xpidlyacc.py \
+  $(DEPTH)/xpcom/idl-parser/cache/xpidllex.py \
+  $(DEPTH)/xpcom/idl-parser/cache/xpidlyacc.py \
   $(NULL)
 
 ifndef MOZ_SYSTEM_PLY
